File: python/transcribe_stream_separate_recognition.py
from typing import Iterable
from soniox.transcribe_live import transcribe_stream
from soniox.speech_service import SpeechClient


def iter_audio() -> Iterable[bytes]:
    with open("../test_data/test_audio_multi_channel.flac", "rb") as fh:
        while True:
            audio = fh.read(1024)
            if len(audio) == 0:
                break
            yield audio


def main():
    with SpeechClient() as client:
        for result in transcribe_stream(
            iter_audio(),
            client,
            model="en_v2_lowlatency",
            include_nonfinal=True,
            num_audio_channels=2,
            enable_separate_recognition_per_channel=True,
        ):
            print(f"Channel {result.channel}: " + "".join(w.text for w in result.words))


if __name__ == "__main__":
    main()

File: node/transcribe_stream_separate_recognition.js
const fs = require("fs");
const { SpeechClient } = require("@soniox/soniox-node");

/**
 * Set your Soniox Cloud API key:
 * from command line: export SONIOX_API_KEY=<YOUR-API-KEY>
 * or
 * pass config object: new SpeechClient({ api_key: "<YOUR-API-KEY>" })
 */
const speechClient = new SpeechClient();

(async function () {
  const onDataHandler = async (result) => {
    console.log(
      `Channel ${result.channel}: ${result.words.map((word) => word.text).join("")}`
    );
  };

  const onEndHandler = (error) => {
    if (error) {
      console.log(`Transcription error: ${error}`);
    }
  };

  // transcribeStream returns object with ".writeAsync()" and ".end()" methods
  // use them to send data and end stream when done.
  const stream = speechClient.transcribeStream(
    {
      model: "en_v2_lowlatency",
      include_nonfinal: true,
      num_audio_channels: 2,
      enable_separate_recognition_per_channel: true,
    },
    onDataHandler,
    onEndHandler
  );

  const CHUNK_SIZE = 1024;
  const streamSource = fs.createReadStream(
    "../test_data/test_audio_multi_channel.flac",
    {
      highWaterMark: CHUNK_SIZE,
    }
  );

  // Simulate data streaming
  for await (const audioChunk of streamSource) {
    await stream.writeAsync(audioChunk);
  }

  stream.end();
})();
